game-core/src/main/java/games/strategy/triplea/ui/mapdata/IslandTerritoryFinder.java

```
@@ -0,0 +1,74 @@
+    final Set<String> seaTerritories = filterSeaTerritories(polygons.keySet());
+    final Set<String> landTerritories = filterNotSeaTerritories(polygons.keySet());
+
+    // map sea territories to: sea territory name -> islands contained by that sea territory
+    return seaTerritories.stream()
+        .collect(
+            Collectors.toMap(sea -> sea, findIslandsForSeaTerritory(landTerritories, polygons)));
+  }
+
+  private static Set<String> filterSeaTerritories(final Set<String> territoryNames) {
+    return filterTerritories(territoryNames, Util::isTerritoryNameIndicatingWater);
+  }
+
+  private static Set<String> filterNotSeaTerritories(final Set<String> territoryNames) {
+    return filterTerritories(territoryNames, Predicate.not(Util::isTerritoryNameIndicatingWater));
+  }
+
+  /** Returns a subset of territories matching a given filter. */
+  private static Set<String> filterTerritories(
+      final Set<String> territoryNames, final Predicate<String> territoryFilter) {
+    return territoryNames.stream().filter(territoryFilter).collect(Collectors.toSet());
+  }
+
+  /** Find all land territories contained by a given sea territory. */
+  private static Function<String, Set<String>> findIslandsForSeaTerritory(
+      final Set<String> landTerritories, final Map<String, List<Polygon>> polygons) {
+
+    return seaTerritory ->
+        landTerritories.stream()
+            .filter(landIsContainedBySeaTerritory(seaTerritory, polygons))
+            .collect(Collectors.toSet());
+  }
+
+  /** Checks that a given land territory would be contained by a given sea territory. */
+  private Predicate<String> landIsContainedBySeaTerritory(
+      final String seaTerritory, final Map<String, List<Polygon>> polygons) {
+
+    // Function where given a territory name, give us a polygon for that territory
+    final Function<String, Polygon> polygonLookup =
+        territoryName -> polygons.get(territoryName).iterator().next();
```

2 things:
1. This is a map of lists, so instead of using `list.iterator().next()`, you could as well use `list.get(0)` directly
2. Does it make sense to check the size of the list before? If the lists was empty for some reason, this would result in an exception. If there are more than 1 polygon (which is mentioned somewhere in the javadocs) this would be silently ignore all the other polygons

---

Original implementation here; I think the polygon reader ensures that we get at least one polygon. I suspect there is further improvement to be had here.